drivers/rtc/rtc_sam0.c

```diff
@@ -0,0 +1,644 @@
+		timeptr->tm_hour, timeptr->tm_min, timeptr->tm_sec);
+
+	return 0;
+}
+
+#ifdef CONFIG_RTC_ALARM
+static uint32_t rtc_sam0_datetime_from_tm(const struct rtc_time *timeptr, uint32_t mask)
+{
+	uint32_t datetime = 0;
+
+	if (mask & RTC_ALARM_TIME_MASK_SECOND) {
+		datetime |= RTC_MODE2_CLOCK_SECOND(timeptr->tm_sec);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MINUTE) {
+		datetime |= RTC_MODE2_CLOCK_MINUTE(timeptr->tm_min);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_HOUR) {
+		datetime |= RTC_MODE2_CLOCK_HOUR(timeptr->tm_hour);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTHDAY) {
+		datetime |= RTC_MODE2_CLOCK_DAY(timeptr->tm_mday);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTH) {
+		datetime |= RTC_MODE2_CLOCK_MONTH(timeptr->tm_mon + 1);
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_YEAR) {
+		datetime |= RTC_MODE2_CLOCK_YEAR(timeptr->tm_year - 99);
+	}
+
+	return datetime;
+}
+
+static void rtc_sam0_tm_from_datetime(struct rtc_time *timeptr, uint32_t mask,
+				      RTC_MODE2_ALARM_Type calendar)
+{
+	memset(timeptr, 0x00, sizeof(struct rtc_time));
+
+	if (mask & RTC_ALARM_TIME_MASK_SECOND) {
+		timeptr->tm_sec = calendar.bit.SECOND;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MINUTE) {
+		timeptr->tm_min = calendar.bit.MINUTE;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_HOUR) {
+		timeptr->tm_hour = calendar.bit.HOUR;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTHDAY) {
+		timeptr->tm_mday = calendar.bit.DAY;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTH) {
+		timeptr->tm_mon = calendar.bit.MONTH - 1;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTH) {
+		timeptr->tm_year = calendar.bit.YEAR + 99;
+	}
+
+	timeptr->tm_wday = -1;
+	timeptr->tm_yday = -1;
+	timeptr->tm_isdst = -1;
+	timeptr->tm_nsec = 0;
+}
+
+static uint32_t rtc_sam0_alarm_msk_from_mask(uint32_t mask)
+{
+	uint32_t alarm_mask = 0;
+
+	if (mask & RTC_ALARM_TIME_MASK_SECOND) {
+		alarm_mask = RTC_MODE2_MASK_SEL_SS_Val;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MINUTE) {
+		alarm_mask = RTC_MODE2_MASK_SEL_MMSS_Val;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_HOUR) {
+		alarm_mask = RTC_MODE2_MASK_SEL_HHMMSS_Val;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTHDAY) {
+		alarm_mask = RTC_MODE2_MASK_SEL_DDHHMMSS_Val;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTH) {
+		alarm_mask = RTC_MODE2_MASK_SEL_MMDDHHMMSS_Val;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_YEAR) {
+		alarm_mask = RTC_MODE2_MASK_SEL_YYMMDDHHMMSS_Val;
+	}
+
+	return alarm_mask;
+}
+
+static uint32_t rtc_sam0_mask_from_alarm_msk(uint32_t alarm_mask)
+{
+	uint32_t mask = 0;
+
+	switch (alarm_mask) {
+	case RTC_MODE2_MASK_SEL_YYMMDDHHMMSS_Val:
+		mask |= RTC_ALARM_TIME_MASK_YEAR;
+	case RTC_MODE2_MASK_SEL_MMDDHHMMSS_Val:
+		mask |= RTC_ALARM_TIME_MASK_MONTH;
+	case RTC_MODE2_MASK_SEL_DDHHMMSS_Val:
+		mask |= RTC_ALARM_TIME_MASK_MONTHDAY;
+	case RTC_MODE2_MASK_SEL_HHMMSS_Val:
+		mask |= RTC_ALARM_TIME_MASK_HOUR;
+	case RTC_MODE2_MASK_SEL_MMSS_Val:
+		mask |= RTC_ALARM_TIME_MASK_MINUTE;
+	case RTC_MODE2_MASK_SEL_SS_Val:
+		mask |= RTC_ALARM_TIME_MASK_SECOND;
+	}
```

Comment:
`default:` is required as per coding guidelines, and should also have explicit `__fallthrough;` statements